The states from the given question



In [28]:
Activity_state = {
    0: 'rest',
    1: 'sweeties',
    2: 'exercises'
}

print(Activity_state)


{0: 'rest', 1: 'sweeties', 2: 'exercises'}


The transition matrix from the given state diagram is given below:

In [29]:
transition_matrix = np.array([[0.2, 0.2, 0.6],
                             [0.2, 0.1, 0.7],
                             [0.1, 0.3, 0.6]])
print(transition_matrix)

[[0.2 0.2 0.6]
 [0.2 0.1 0.7]
 [0.1 0.3 0.6]]


The random walk could be as follow:

In [30]:
import numpy as np

# Defining the transition matrix
transition_matrix = np.array([[0.7, 0.2, 0.1],
                              [0.3, 0.6, 0.1],
                              [0.1, 0.4, 0.5]])

# Defining the activity states
Activity_state = {
    0: 'rest',
    1: 'sweeties',
    2: 'exercises'
}

# Setting up the number of steps and starting state
n = 15
starting_state = 0
previous_state = starting_state

# Now performimg the random walk
for i in range(n):
    current_state = np.random.choice([0, 1, 2], p=transition_matrix[previous_state])
    print(Activity_state[current_state], "->", end="")
    # Update the previous state
   


    

sweeties ->rest ->rest ->rest ->rest ->rest ->rest ->rest ->rest ->sweeties ->sweeties ->rest ->rest ->exercises ->sweeties ->

The starting state is Rest. Calculate the probability of possible activity on the 15th day.
What about the state and probability after 10000 days?

In [31]:
import numpy as np

# Transition matrix
transition_matrix = np.array([[0.2, 0.2, 0.6],
                             [0.2, 0.1, 0.7],
                             [0.1, 0.3, 0.6]])
# Starting state vector
start_state = np.array([1, 0, 0])  # Rest is the starting state
probability_15_days = np.linalg.matrix_power(transition_matrix.T, 15) @ start_state # Calculating probability after 15 days
probability_10000_days = np.linalg.matrix_power(transition_matrix.T, 10000) @ start_state # Calculating probability after 10,000 days
print("Probability of activities on the 15th day:") # output of the respective probabilities
print("Rest:", probability_15_days[0])
print("Sweeties:",probability_15_days[1])
print("Exercises:", probability_15_days[2])

print("\nProbability of activities after 10,000 days:")
print("Rest:", probability_10000_days[0])
print("Sweeties:", probability_10000_days[1])
print("Exercises:", probability_10000_days[2])


Probability of activities on the 15th day:
Rest: 0.137614678899138
Sweeties: 0.238532110091598
Exercises: 0.6238532110092639

Probability of activities after 10,000 days:
Rest: 0.13761467889908133
Sweeties: 0.23853211009174094
Exercises: 0.6238532110091686


What do you observe from the above two?

While solving the above probability, we get to know  that as the number of days went up, the odds tend to stay the same. Even after a long time, 10,000 days case, the chances  in each state get closeby to certain values. Here,in this case, the chances after 10,000 days are very close to the chances after 15 days. It represents that the distribution reached a stable state. We received information that as the number of steps goes up, the probability values get closer together. Where, if the number of trials is low, the model supposed to give too high of odds for rest and sweeties. Hence, as the number of trials goes up, the activities after 10,000 days  settle on 13.76% for rest, 23.85% for sweets, and 62.39% for exercises.